File: algorithm/python/SplitLinkedListInParts.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def splitListToParts(self, root, k):
        length, node = 0, root
        
        while node:
            length += 1
            node = node.next
        
        n, remaining = divmod(length, k)
        
        # split the linked list
        res, curr = [], root
        
        for i in range(k):
            head = curr
            for _ in range(n + (i < remaining) - 1):
                if curr: curr = curr.next
            if curr:
                curr.next, curr = None, curr.next
            res.append(head)
        return res

File: algorithm/JavaScript/SplitLinkedListInParts.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode[]}
 */
 var splitListToParts = function(head, k) {
  let length = 0, curr = head;
  while (curr) {
      length += 1;
      curr = curr.next;
  }
  
  let n = Math.floor(length / k), remaining = length % k;
  
  const res = []; 
  curr = head;
  
  for (let i = 0; i < k; i++) {
      const node = curr;
      const width = n + (i < remaining ? 1 : 0);
      for (let j = 1; j < width; j++) {
          if (curr) curr = curr.next;
      }
      if (curr) {
          const next = curr.next;
          curr.next = null;
          curr = next;
      }
      res.push(node);
  }
  
  return res
  
};